contrib/firebaseio/src/test/java/io/BaseFirebaseSourceTest.java

```diff
@@ -0,0 +1,168 @@
+import com.google.cloud.dataflow.sdk.values.PCollection;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.firebase.client.ChildEventListener;
+import com.firebase.client.Firebase;
+import com.firebase.client.Firebase.CompletionListener;
+import com.firebase.client.FirebaseError;
+import com.firebase.client.ValueEventListener;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import contrib.LogElements;
+
+import events.FirebaseEvent;
+import utils.FirebaseAuthenticator;
+import utils.FirebaseEmptyAuthenticator;
+import utils.LoggingArrayList;
+
+
+
+/**
+ * Archetype for comparing events produced by through {@link FirebaseSource} with events surfaced
+ * by the unmodified {@link ChildEventListener} and {@link ValueEventListener}.
+ **/
+@RunWith(JUnit4.class)
+public abstract class BaseFirebaseSourceTest {
+
+  protected FirebaseAuthenticator auther = new FirebaseEmptyAuthenticator();
+  private Throwable err;
+  protected Firebase testRef;
+
+  static final Logger LOGGER = LoggerFactory.getLogger(BaseFirebaseSourceTest.class);
+
+  final LoggingArrayList<FirebaseEvent<JsonNode>> expected =
+      new LoggingArrayList<FirebaseEvent<JsonNode>>(LOGGER);
+
+  @Before
+  public void setUp() throws JsonParseException,
+  JsonMappingException, IOException, InterruptedException {
+
+    testRef = new Firebase("https://dataflowio.firebaseio-demo.com")
+        .child(Integer.toHexString(this.hashCode())
+            + Long.toHexString(System.currentTimeMillis()));
+
+    cleanFirebase(testRef);
+
+    prepareData(new ObjectMapper().<List<Map<String, Object>>>readValue(
+        ClassLoader.getSystemResourceAsStream("testdata.json"),
+        new TypeReference<List<Object>>() {}));
+
+  }
+
+  @After
+  public void teardown() throws InterruptedException{
+    cleanFirebase(testRef);
+  }
+
+  protected Thread getPipelineThread(final Pipeline p){
+    Thread t = new Thread(new Runnable(){
+      @Override
+      public void run() {
+        LOGGER.info(p.run().getState().toString());
+      }
+    });
+    t.setUncaughtExceptionHandler(new UncaughtExceptionHandler(){
+      @Override
+      public void uncaughtException(Thread t, Throwable e) {
+        err = e;
+      }
+    });
+    return t;
+  }
+
+  protected void cleanFirebase(Firebase f) throws InterruptedException{
+    final Semaphore lock = new Semaphore(1);
```

Comment:
Your uses of semaphores might be cleaned up with [CountdownLatch](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html). I think the way you're using it to synchronize is effectively the "sample usage" in the main Javadoc